packages/button/components/ButtonBase.tsx

```diff
@@ -0,0 +1,164 @@
+import * as React from "react";
+import { cx } from "emotion";
+import {
+  buttonReset,
+  textWeight,
+  flex,
+  flexItem,
+  display,
+  padding
+} from "../../shared/styles/styleUtils";
+import {
+  button,
+  buttonBase,
+  buttonContent,
+  buttonInverse,
+  fullWidthButton,
+  getInverseMutedButtonStyles,
+  getMutedButtonStyles,
+  processingText
+} from "../style";
+
+export enum ButtonAppearances {
+  Primary = "primary",
+  Secondary = "secondary",
+  Standard = "standard",
+  Danger = "danger",
+  Success = "success"
+}
+
+export interface ButtonProps {
+  /**
+   * if the button triggers new content to appear (e.g.: modals and dropdowns)
+   */
+  ariaHaspopup?: boolean;
+  children?: React.ReactNode | string;
+  /**
+   * whether or not the button is enabled
+   */
+  disabled?: boolean;
+  /**
+   * the icon that appears before the button text
+   */
+  iconStart?: React.ReactElement<HTMLElement>;
+  /**
+   * the icon that appears after the button text
+   */
+  iconEnd?: React.ReactElement<HTMLElement>;
+  /**
+   * whether the action the button was meant to do has completed
+   */
+  isProcessing?: boolean;
+  /**
+   * if the button should fill the width of it's container
+   */
+  isFullWidth?: boolean;
+  /**
+   * if the button is on a dark background
+   */
+  isInverse?: boolean;
+  /**
+   * the function the button calls
+   */
+  onClick?: (e?: React.SyntheticEvent<HTMLElement>) => void;
+  /**
+   * the type of button - same as HTML "type" attribute on the <button> tag
+   */
+  type?: "button" | "reset" | "submit";
+}
+
+export interface ButtonBaseProps extends ButtonProps {
+  appearance: ButtonAppearances;
+  className?: string;
+}
+
+class ButtonBase extends React.PureComponent<ButtonBaseProps, {}> {
+  constructor(props) {
+    super(props);
+    this.onClick = this.onClick.bind(this);
+  }
+  public render() {
+    const {
+      appearance,
+      children,
+      className,
+      disabled,
+      iconStart,
+      iconEnd,
+      isInverse,
+      isProcessing,
+      isFullWidth,
+      onClick,
+      ...other
```

Comment:
This long props list makes me feel like this component might be to configurable instead of being composeable. For example `iconStart` and `iconEnd` why do we need these as props couldn't we just provide the icons with the Children like this:
```JS
<Button>
  <Icon name="xyz"/>
  Loading
  <Icon name="end" />
</Button>
```
And then check the Button children for Icons component if we actually need different styling for these. 

Why do we need the `isInverse` prop? isn't that just another appearance?

Reply:
We _could_ just require they get passed in as children, but there's some layout and spacing styles attached to the divs that wrap icons, and I don't want engineers to have to worry about applying those styles every time they need to use a button with an icon

Reply:
The appearances have meaning attached to them (see the README). `isInverse` is just a way to adapt any button appearance to a dark background.